File: python/tests/test_entity_normalizer.py
"""Tests for entity_normalizer — port of store/entity-normalizer.js.

Covers each normalize_* helper plus normalize_raw against minimal,
legacy-shaped, and fully-populated inputs.
"""

from kalairos.entity_normalizer import (
    _SOURCE_TRUST_DEFAULTS,
    _VALID_MEMORY_TYPES,
    default_trust_score,
    make_version_id,
    normalize_classification,
    normalize_memory_type,
    normalize_raw,
    normalize_retention,
    normalize_workspace_id,
)
from kalairos.versioning import ACTIONS, is_valid_action, synthesize_action


# ── default_trust_score ────────────────────────────────────────────────────


def test_default_trust_score_known_types():
    assert default_trust_score("user") == 0.90
    assert default_trust_score("agent") == 0.75
    assert default_trust_score("tool") == 0.80
    assert default_trust_score("file") == 0.70
    assert default_trust_score("system") == 0.60


def test_default_trust_score_unknown_falls_back():
    assert default_trust_score("unknown") == 0.70
    assert default_trust_score("") == 0.70
    assert default_trust_score(None) == 0.70


# ── normalize_classification ───────────────────────────────────────────────


def test_normalize_classification_defaults_to_internal():
    assert normalize_classification(None) == "internal"
    assert normalize_classification("") == "internal"
    assert normalize_classification("   ") == "internal"


def test_normalize_classification_canonicalizes():
    assert normalize_classification("Public") == "public"
    assert normalize_classification("  Confidential  ") == "confidential"


def test_normalize_classification_preserves_custom_strings():
    """Caller-defined policy labels survive normalization (lowercased)."""
    assert normalize_classification("hr-only") == "hr-only"
    assert normalize_classification("PII-Locked") == "pii-locked"


# ── normalize_retention ────────────────────────────────────────────────────


def test_normalize_retention_defaults():
    assert normalize_retention(None) == {"policy": "keep", "expiresAt": None}
    assert normalize_retention({}) == {"policy": "keep", "expiresAt": None}
    assert normalize_retention("not a dict") == {"policy": "keep", "expiresAt": None}


def test_normalize_retention_keeps_finite_expires_at():
    r = normalize_retention({"policy": "delete", "expiresAt": 1234567890})
    assert r == {"policy": "delete", "expiresAt": 1234567890}


def test_normalize_retention_drops_non_finite_expires_at():
    assert normalize_retention({"policy": "x", "expiresAt": float("inf")})["expiresAt"] is None
    assert normalize_retention({"policy": "x", "expiresAt": float("nan")})["expiresAt"] is None
    assert normalize_retention({"policy": "x", "expiresAt": "1234"})["expiresAt"] is None


def test_normalize_retention_rejects_bool_expires_at():
    """Python bools subclass int — JS Number.isFinite rejects bools, so
    we should too."""
    assert normalize_retention({"policy": "x", "expiresAt": True})["expiresAt"] is None


# ── normalize_memory_type ──────────────────────────────────────────────────


def test_normalize_memory_type_valid_values():
    for v in ("short-term", "long-term", "working"):
        assert normalize_memory_type(v) == v


def test_normalize_memory_type_invalid_falls_back_to_long_term():
    assert normalize_memory_type("episodic") == "long-term"


def test_normalize_memory_type_defaults_and_canonicalizes():
    assert normalize_memory_type(None) == "long-term"
    assert normalize_memory_type("") == "long-term"
    assert normalize_memory_type("  Working  ") == "working"


# ── normalize_workspace_id ─────────────────────────────────────────────────


def test_normalize_workspace_id_defaults():
    assert normalize_workspace_id(None) == "default"
    assert normalize_workspace_id("") == "default"
    assert normalize_workspace_id("   ") == "default"


def test_normalize_workspace_id_passes_through():
    assert normalize_workspace_id("alpha") == "alpha"
    assert normalize_workspace_id("  alpha-1  ") == "alpha-1"


# ── make_version_id ────────────────────────────────────────────────────────


def test_make_version_id_format():
    assert make_version_id("ent-42", 1) == "ent-42:1"
    assert make_version_id("ent-42", 7) == "ent-42:7"


# ── synthesize_action / is_valid_action ────────────────────────────────────


def test_synthesize_action_first_is_remembered():
    assert synthesize_action(0, None) == ACTIONS["remembered"]
    assert synthesize_action(0, {"type": "correction"}) == ACTIONS["remembered"]


def test_synthesize_action_correction():
    assert synthesize_action(1, {"type": "correction"}) == ACTIONS["corrected"]


def test_synthesize_action_consolidation_and_default():
    assert synthesize_action(1, {"type": "consolidation"}) == ACTIONS["superseded"]
    assert synthesize_action(1, None) == ACTIONS["superseded"]
    assert synthesize_action(1, {"type": "other"}) == ACTIONS["superseded"]


def test_is_valid_action():
    for v in ACTIONS.values():
        assert is_valid_action(v)
    assert not is_valid_action("nope")
    assert not is_valid_action(None)
    assert not is_valid_action(42)


# ── normalize_raw ──────────────────────────────────────────────────────────


def test_normalize_raw_fills_minimal_entity():
    """An entity with only id + text comes out with all defaults populated."""
    raw = {"id": "ent-1", "text": "hello"}
    out = normalize_raw(raw)
    assert out is raw  # mutated in place
    assert out["links"] == []
    assert out["versions"] == []
    assert out["type"] == "text"
    assert out["metadata"] == {}
    assert out["tags"] == []
    assert out["classification"] == "internal"
    assert out["retention"] == {"policy": "keep", "expiresAt": None}
    assert out["memoryType"] == "long-term"
    assert out["workspaceId"] == "default"
    assert out["importance"] is None
    assert out["trustScore"] == 0.90  # default for source {"type": "user"}
    assert out["deletedAt"] is None
    assert out["deletedBy"] is None
    assert out["llmKeywords"] == []
    assert out["trailEvents"] == []


def test_normalize_raw_preserves_explicit_fields():
    raw = {
        "id": "ent-2",
        "text": "x",
        "memoryType": "working",
        "workspaceId": "alpha",
        "trustScore": 0.42,
        "importance": 5,
        "tags": ["a", "b"],
    }
    out = normalize_raw(raw)
    assert out["memoryType"] == "working"
    assert out["workspaceId"] == "alpha"
    assert out["trustScore"] == 0.42
    assert out["importance"] == 5
    assert out["tags"] == ["a", "b"]


def test_normalize_raw_links_become_list():
    raw = {"id": "x", "text": "x", "links": ["a", "b", "c"]}
    out = normalize_raw(raw)
    assert out["links"] == ["a", "b", "c"]


def test_normalize_raw_soft_delete_fields():
    raw = {"id": "x", "text": "x", "deletedAt": 1234, "deletedBy": "user-1"}
    out = normalize_raw(raw)
    assert out["deletedAt"] == 1234
    assert out["deletedBy"] == "user-1"


def test_normalize_raw_default_trust_from_source():
    """trustScore default depends on source.type."""
    raw = {"id": "x", "text": "x", "source": {"type": "agent"}}
    out = normalize_raw(raw)
    assert out["trustScore"] == 0.75


def test_normalize_raw_legacy_versions_oldest_first_get_reversed():
    """JSONL with versions sorted oldest→newest must be flipped to
    newest-first to match the in-memory invariant."""
    raw = {
        "id": "ent",
        "text": "v3",
        "versions": [
            {"timestamp": 100, "text": "v1"},
            {"timestamp": 200, "text": "v2"},
            {"timestamp": 300, "text": "v3"},
        ],
    }
    out = normalize_raw(raw)
    assert [v["timestamp"] for v in out["versions"]] == [300, 200, 100]


def test_normalize_raw_synthesises_audit_fields():
    """Legacy versions missing audit fields get them filled deterministically."""
    raw = {
        "id": "ent",
        "text": "v2",
        "versions": [
            {"timestamp": 200, "text": "v2"},  # newest-first
            {"timestamp": 100, "text": "v1"},
        ],
    }
    out = normalize_raw(raw)
    v_new, v_old = out["versions"][0], out["versions"][1]
    assert v_old["versionId"] == "ent:1"
    assert v_new["versionId"] == "ent:2"
    assert v_old["ingestAt"] == 100
    assert v_old["effectiveAt"] == 100
    assert v_old["validFrom"] == 100
    # validTo: oldest closes at next version's effectiveAt; newest is open.
    assert v_old["validTo"] == 200
    assert v_new["validTo"] is None
    assert v_old["previousVersionId"] is None
    assert v_new["previousVersionId"] == "ent:1"
    assert v_old["who"] is None and v_old["why"] is None
    assert v_old["action"] == ACTIONS["remembered"]
    assert v_new["action"] == ACTIONS["superseded"]


def test_normalize_raw_soft_delete_closes_latest_version():
    raw = {
        "id": "ent",
        "text": "v1",
        "deletedAt": 999,
        "versions": [{"timestamp": 100, "text": "v1"}],
    }
    out = normalize_raw(raw)
    assert out["versions"][0]["validTo"] == 999


def test_normalize_raw_per_version_classification_inherits():
    raw = {
        "id": "ent",
        "text": "x",
        "classification": "Confidential",
        "versions": [{"timestamp": 1, "text": "x"}],
    }
    out = normalize_raw(raw)
    assert out["classification"] == "confidential"
    assert out["versions"][0]["classification"] == "confidential"


def test_normalize_raw_llm_keywords_from_metadata():
    raw = {
        "id": "ent",
        "text": "x",
        "metadata": {"llm": {"keywords": ["alpha", "beta"]}},
    }
    out = normalize_raw(raw)
    assert out["llmKeywords"] == ["alpha", "beta"]


def test_normalize_raw_preserves_existing_audit_fields():
    """If a version already carries audit fields, normalize must not
    overwrite them."""
    raw = {
        "id": "ent",
        "text": "v1",
        "versions": [
            {
                "timestamp": 100,
                "text": "v1",
                "versionId": "custom-id",
                "action": "imported",
                "who": "alice",
                "why": "bulk import",
                "ingestAt": 50,
                "effectiveAt": 75,
                "validFrom": 75,
                "validTo": 200,
                "previousVersionId": "prev",
            }
        ],
    }
    out = normalize_raw(raw)
    v = out["versions"][0]
    assert v["versionId"] == "custom-id"
    assert v["action"] == "imported"
    assert v["who"] == "alice"
    assert v["why"] == "bulk import"
    assert v["ingestAt"] == 50
    assert v["effectiveAt"] == 75
    assert v["validFrom"] == 75
    assert v["validTo"] == 200
    assert v["previousVersionId"] == "prev"


def test_normalize_raw_idempotent():
    """Running normalize_raw twice on the same dict must produce the same
    result — important when the same JSONL row is replayed during a
    rebuild or replay-forward."""
    raw = {"id": "ent", "text": "x", "versions": [{"timestamp": 1, "text": "x"}]}
    once = normalize_raw(raw)
    snapshot = dict(once)
    snapshot["versions"] = [dict(v) for v in once["versions"]]
    twice = normalize_raw(once)
    assert twice["versions"][0]["versionId"] == snapshot["versions"][0]["versionId"]
    assert twice["versions"][0]["action"] == snapshot["versions"][0]["action"]
    assert twice["trustScore"] == snapshot["trustScore"]

File: tests/test-entity-normalizer.js
// test-entity-normalizer.js — store/entity-normalizer.js unit tests
// Run: node tests/test-entity-normalizer.js
"use strict";

const assert = require("assert/strict");

const {
  defaultTrustScore,
  normalizeClassification,
  normalizeRetention,
  normalizeMemoryType,
  normalizeWorkspaceId,
  makeVersionId,
  normalizeRaw,
} = require("../store/entity-normalizer");

// ─── Minimal test runner (mirrors test-basic.js / test-sqlite-index.js) ───────
let passed = 0, failed = 0;
async function test(name, fn) {
  try {
    await fn();
    console.log(`  ✅  ${name}`);
    passed++;
  } catch (e) {
    console.log(`  ❌  ${name}`);
    console.log(`       ${e.message}`);
    failed++;
  }
}

(async () => {

// ── defaultTrustScore ───────────────────────────────────────────────────────
console.log("\n── defaultTrustScore ─────────────────────────────────────────────");

await test("returns the documented default for each known source type", () => {
  assert.equal(defaultTrustScore("user"),   0.90);
  assert.equal(defaultTrustScore("agent"),  0.75);
  assert.equal(defaultTrustScore("tool"),   0.80);
  assert.equal(defaultTrustScore("file"),   0.70);
  assert.equal(defaultTrustScore("system"), 0.60);
});

await test("falls back to 0.70 for an unknown source type", () => {
  assert.equal(defaultTrustScore("bogus"), 0.70);
});

await test("falls back to 0.70 for undefined/empty input", () => {
  assert.equal(defaultTrustScore(undefined), 0.70);
  assert.equal(defaultTrustScore(""), 0.70);
});

// ── normalizeClassification ─────────────────────────────────────────────────
console.log("\n── normalizeClassification ───────────────────────────────────────");

await test("empty/falsy input defaults to 'internal'", () => {
  assert.equal(normalizeClassification(""), "internal");
  assert.equal(normalizeClassification(null), "internal");
  assert.equal(normalizeClassification(undefined), "internal");
  assert.equal(normalizeClassification(0), "internal");
});

await test("whitespace-only input defaults to 'internal'", () => {
  assert.equal(normalizeClassification("   "), "internal");
  assert.equal(normalizeClassification("\t\n"), "internal");
});

await test("trims and lowercases mixed-case input", () => {
  assert.equal(normalizeClassification("  Confidential  "), "confidential");
  assert.equal(normalizeClassification("REGULATED"), "regulated");
});

await test("preserves unicode content through trim/lowercase", () => {
  assert.equal(normalizeClassification("  Confidential — 機密  "), "confidential — 機密");
});

// ── normalizeRetention ──────────────────────────────────────────────────────
console.log("\n── normalizeRetention ────────────────────────────────────────────");

await test("null/undefined default to { policy: 'keep', expiresAt: null }", () => {
  assert.deepEqual(normalizeRetention(null),      { policy: "keep", expiresAt: null });
  assert.deepEqual(normalizeRetention(undefined), { policy: "keep", expiresAt: null });
});

await test("non-object input (string, number) also defaults", () => {
  assert.deepEqual(normalizeRetention("purge"), { policy: "keep", expiresAt: null });
  assert.deepEqual(normalizeRetention(42),      { policy: "keep", expiresAt: null });
});

await test("an array is typeof 'object' and passes through as a default-shaped policy", () => {
  // Arrays are truthy and typeof "object", so they take the object branch;
  // array.policy/.expiresAt are both undefined, so the result is still the
  // default shape. Documents the actual (slightly surprising) behavior.
  assert.deepEqual(normalizeRetention([1, 2, 3]), { policy: "keep", expiresAt: null });
});

await test("normalizes policy casing/whitespace and preserves a finite expiresAt", () => {
  assert.deepEqual(
    normalizeRetention({ policy: "  Purge  ", expiresAt: 123456 }),
    { policy: "purge", expiresAt: 123456 }
  );
});

await test("non-finite expiresAt (NaN, Infinity, missing) normalizes to null", () => {
  assert.equal(normalizeRetention({ expiresAt: NaN }).expiresAt, null);
  assert.equal(normalizeRetention({ expiresAt: Infinity }).expiresAt, null);
  assert.equal(normalizeRetention({}).expiresAt, null);
});

// ── normalizeMemoryType ─────────────────────────────────────────────────────
console.log("\n── normalizeMemoryType ───────────────────────────────────────────");

await test("accepts each of the three valid memory types", () => {
  assert.equal(normalizeMemoryType("short-term"), "short-term");
  assert.equal(normalizeMemoryType("long-term"),  "long-term");
  assert.equal(normalizeMemoryType("working"),    "working");
});

await test("is case/whitespace-insensitive for valid types", () => {
  assert.equal(normalizeMemoryType("  Working  "), "working");
  assert.equal(normalizeMemoryType("SHORT-TERM"), "short-term");
});

await test("falls back to 'long-term' for empty or unrecognized input", () => {
  assert.equal(normalizeMemoryType(""), "long-term");
  assert.equal(normalizeMemoryType(null), "long-term");
  assert.equal(normalizeMemoryType("bogus"), "long-term");
});

// ── normalizeWorkspaceId ────────────────────────────────────────────────────
console.log("\n── normalizeWorkspaceId ──────────────────────────────────────────");

await test("empty/falsy input defaults to 'default'", () => {
  assert.equal(normalizeWorkspaceId(""), "default");
  assert.equal(normalizeWorkspaceId(null), "default");
  assert.equal(normalizeWorkspaceId(undefined), "default");
});

await test("whitespace-only input defaults to 'default'", () => {
  assert.equal(normalizeWorkspaceId("   "), "default");
});

await test("trims but does not lowercase (workspace IDs are case-sensitive)", () => {
  assert.equal(normalizeWorkspaceId("  Team-Legal  "), "Team-Legal");
});

await test("preserves unicode workspace IDs", () => {
  assert.equal(normalizeWorkspaceId("  日本語ワークスペース  "), "日本語ワークスペース");
});

// ── makeVersionId ────────────────────────────────────────────────────────────
console.log("\n── makeVersionId ─────────────────────────────────────────────────");

await test("formats as `${entityId}:${ordinal}`", () => {
  assert.equal(makeVersionId(123, 1), "123:1");
  assert.equal(makeVersionId("abc", 2), "abc:2");
});

await test("is deterministic — same input always produces the same output", () => {
  const calls = Array.from({ length: 20 }, () => makeVersionId(42, 7));
  assert.ok(calls.every(v => v === "42:7"), "every call with identical args must match");
});

await test("distinguishes ordinals and entity ids independently", () => {
  assert.notEqual(makeVersionId(1, 1), makeVersionId(1, 2));
  assert.notEqual(makeVersionId(1, 1), makeVersionId(2, 1));
});

await test("handles ordinal 0 and unicode entity ids without special-casing", () => {
  assert.equal(makeVersionId(5, 0), "5:0");
  assert.equal(makeVersionId("実体-7", 3), "実体-7:3");
});

// ── normalizeRaw — bare defaults ────────────────────────────────────────────
console.log("\n── normalizeRaw: defaults on minimal/empty input ─────────────────");

await test("an entirely empty object gets every default field", () => {
  const out = normalizeRaw({});
  assert.equal(out.type, "text");
  assert.deepEqual(out.metadata, {});
  assert.deepEqual(out.tags, []);
  assert.ok(out.links instanceof Set);
  assert.equal(out.links.size, 0);
  assert.deepEqual(out.versions, []);
  assert.deepEqual(out.source, { type: "user" });
  assert.equal(out.classification, "internal");
  assert.deepEqual(out.retention, { policy: "keep", expiresAt: null });
  assert.equal(out.deletedAt, null);
  assert.equal(out.deletedBy, null);
  assert.equal(out.memoryType, "long-term");
  assert.equal(out.workspaceId, "default");
  assert.deepEqual(out.llmKeywords, []);
  assert.equal(out.importance, null);
  assert.equal(out.trustScore, 0.90); // default source is "user" → 0.90
  assert.equal(out.contentRisk, null);
  assert.deepEqual(out.trailEvents, []);
});

await test("mutates the input object in place and returns the same reference", () => {
  const raw = { id: 1, text: "hello" };
  const out = normalizeRaw(raw);
  assert.equal(out, raw, "normalizeRaw must return the same object it was given");
  assert.equal(raw.type, "text", "the original object should itself be mutated");
});

await test("existing links array is converted to a deduplicated Set", () => {
  const out = normalizeRaw({ id: 1, links: [1, 1, 2, 3, 3, 3] });
  assert.deepEqual([...out.links].sort(), [1, 2, 3]);
});

// ── normalizeRaw — blank/whitespace text and tags ───────────────────────────
console.log("\n── normalizeRaw: blank input on optional fields ──────────────────");

await test("empty string text is left as-is (not replaced with a default)", () => {
  // Only `type`/`tags`/`metadata`/`links`/`versions` get falsy-based defaults;
  // `text` itself passes through untouched.
  const out = normalizeRaw({ id: 1, text: "" });
  assert.equal(out.text, "");
});

await test("empty tags array stays empty; falsy tags become []", () => {
  assert.deepEqual(normalizeRaw({ id: 1, tags: [] }).tags, []);
  assert.deepEqual(normalizeRaw({ id: 1, tags: null }).tags, []);
});

// ── normalizeRaw — zero is not "missing" ────────────────────────────────────
console.log("\n── normalizeRaw: 0 is a real value, not treated as absent ────────");

await test("explicit 0 for contentRisk/importance/trustScore is preserved, not defaulted", () => {
  const out = normalizeRaw({ id: 1, versions: [], contentRisk: 0, importance: 0, trustScore: 0 });
  assert.equal(out.contentRisk, 0, "contentRisk=0 (assessed-clean) must not become null");
  assert.equal(out.importance, 0);
  assert.equal(out.trustScore, 0);
});

await test("missing contentRisk/importance backfill to null, not 0", () => {
  const out = normalizeRaw({ id: 1, versions: [] });
  assert.equal(out.contentRisk, null);
  assert.equal(out.importance, null);
});

// ── normalizeRaw — llmKeywords ──────────────────────────────────────────────
console.log("\n── normalizeRaw: llmKeywords backfill ─────────────────────────────");

await test("pulls llmKeywords from metadata.llm.keywords when missing", () => {
  const out = normalizeRaw({ id: 1, versions: [], metadata: { llm: { keywords: ["alpha", "beta"] } } });
  assert.deepEqual(out.llmKeywords, ["alpha", "beta"]);
});

await test("defaults llmKeywords to [] when neither field is present", () => {
  const out = normalizeRaw({ id: 1, versions: [] });
  assert.deepEqual(out.llmKeywords, []);
});

await test("an existing llmKeywords array is left untouched, even if metadata disagrees", () => {
  const out = normalizeRaw({
    id: 1, versions: [],
    llmKeywords: ["preset"],
    metadata: { llm: { keywords: ["other"] } },
  });
  assert.deepEqual(out.llmKeywords, ["preset"]);
});

// ── normalizeRaw — source/classification inheritance from versions ─────────
console.log("\n── normalizeRaw: source/classification inherited from versions ───");

await test("raw.source/classification fall back to the first version that carries them", () => {
  const out = normalizeRaw({
    id: 1,
    versions: [
      { timestamp: 200 },
      { timestamp: 100, source: { type: "file" }, classification: "confidential" },
    ],
  });
  assert.deepEqual(out.source, { type: "file" });
  assert.equal(out.classification, "confidential");
  // trustScore should follow the inherited source type's default.
  assert.equal(out.trustScore, defaultTrustScore("file"));
});

await test("every version without its own source inherits raw.source after normalization", () => {
  const out = normalizeRaw({
    id: 1,
    source: { type: "tool" },
    versions: [{ timestamp: 1 }, { timestamp: 2 }],
  });
  for (const v of out.versions) assert.deepEqual(v.source, { type: "tool" });
});

// ── normalizeRaw — deletedAt / soft delete ──────────────────────────────────
console.log("\n── normalizeRaw: soft-delete (deletedAt) handling ─────────────────");

await test("no deletedAt → deletedAt/deletedBy both stay null", () => {
  const out = normalizeRaw({ id: 1, versions: [] });
  assert.equal(out.deletedAt, null);
  assert.equal(out.deletedBy, null);
});

await test("deletedAt is coerced to a Number and deletedBy defaults to null", () => {
  const out = normalizeRaw({ id: 1, versions: [{ timestamp: 1 }], deletedAt: "555" });
  assert.equal(out.deletedAt, 555);
  assert.equal(out.deletedBy, null);
});

await test("a positive deletedAt closes the latest version's open validTo", () => {
  const out = normalizeRaw({ id: 1, versions: [{ timestamp: 500 }], deletedAt: 999 });
  assert.equal(out.versions[0].validTo, 999);
});

await test("deletedAt does not override a validTo the version already had", () => {
  const out = normalizeRaw({
    id: 1,
    versions: [{ timestamp: 500, validTo: 700 }],
    deletedAt: 999,
  });
  assert.equal(out.versions[0].validTo, 700, "explicit validTo must win over deletedAt backfill");
});

await test("BUGFIX: deletedAt=0 (a legitimate epoch timestamp) still closes validTo", () => {
  // Regression test for a truthy-check bug: `if (raw.deletedAt && ...)` treated
  // deletedAt=0 as "not deleted" and left validTo open. Fixed to `!== null`.
  const out = normalizeRaw({ id: 1, versions: [{ timestamp: 100 }], deletedAt: 0 });
  assert.equal(out.deletedAt, 0, "deletedAt itself must be preserved as 0, not defaulted");
  assert.equal(out.versions[0].validTo, 0, "validTo must be closed at deletedAt=0, not left open");
});

// ── normalizeRaw — version chain backfill (single version) ─────────────────
console.log("\n── normalizeRaw: version chain backfill (single version) ─────────");

await test("a lone legacy version gets versionId/ingestAt/effectiveAt/validFrom/action synthesized", () => {
  const out = normalizeRaw({ id: 77, versions: [{ timestamp: 1000 }] });
  const v = out.versions[0];
  assert.equal(v.versionId, "77:1");
  assert.equal(v.ingestAt, 1000);
  assert.equal(v.effectiveAt, 1000);
  assert.equal(v.validFrom, 1000);
  assert.equal(v.validTo, null, "the only (latest) version stays open-ended");
  assert.equal(v.previousVersionId, null);
  assert.equal(v.who, null);
  assert.equal(v.why, null);
  assert.equal(v.action, "remembered", "first-ever version synthesizes to 'remembered'");
});

// ── normalizeRaw — version chain backfill (multi-version) ──────────────────
console.log("\n── normalizeRaw: version chain backfill (multi-version) ──────────");

await test("a 3-version chain gets correctly linked previousVersionId and closed validTo", () => {
  const out = normalizeRaw({
    id: 20,
    versions: [
      { timestamp: 300 }, // newest
      { timestamp: 200 },
      { timestamp: 100 }, // oldest
    ],
  });
  const [v3, v2, v1] = out.versions; // still newest-first in the output array

  assert.equal(v1.versionId, "20:1");
  assert.equal(v2.versionId, "20:2");
  assert.equal(v3.versionId, "20:3");

  assert.equal(v1.previousVersionId, null);
  assert.equal(v2.previousVersionId, "20:1");
  assert.equal(v3.previousVersionId, "20:2");

  // Each closed version's validTo is the next version's effectiveAt.
  assert.equal(v1.validTo, 200);
  assert.equal(v2.validTo, 300);
  assert.equal(v3.validTo, null, "latest version stays open");

  assert.equal(v1.action, "remembered");
  assert.equal(v2.action, "superseded");
  assert.equal(v3.action, "superseded");
});

await test("validTo backfill uses the next version's effectiveAt, not its timestamp, when they differ", () => {
  const out = normalizeRaw({
    id: 21,
    versions: [
      { timestamp: 300, effectiveAt: 310 }, // newest — explicit effectiveAt differs from timestamp
      { timestamp: 200 },                    // oldest
    ],
  });
  const [, vOldest] = out.versions; // still newest-first: index 1 is the older one
  assert.equal(vOldest.validTo, 310, "validTo must follow the closing version's effectiveAt, not its timestamp");
});

await test("a delta.type of 'correction' synthesizes the 'corrected' action", () => {
  const out = normalizeRaw({
    id: 22,
    versions: [
      { timestamp: 200, delta: { type: "correction" } },
      { timestamp: 100 },
    ],
  });
  assert.equal(out.versions[0].action, "corrected");
  assert.equal(out.versions[1].action, "remembered");
});

await test("existing fields on a version are never overwritten by the backfill", () => {
  const out = normalizeRaw({
    id: 23,
    versions: [{
      timestamp: 300,
      versionId: "custom:1",
      ingestAt: 1,
      effectiveAt: 2,
      validFrom: 3,
      validTo: 4,
      previousVersionId: "prev-custom",
      who: "alice",
      why: "manual correction",
      action: "contested",
    }],
  });
  const v = out.versions[0];
  assert.equal(v.versionId, "custom:1");
  assert.equal(v.ingestAt, 1);
  assert.equal(v.effectiveAt, 2);
  assert.equal(v.validFrom, 3);
  assert.equal(v.validTo, 4);
  assert.equal(v.previousVersionId, "prev-custom");
  assert.equal(v.who, "alice");
  assert.equal(v.why, "manual correction");
  assert.equal(v.action, "contested");
});

await test("linkIds defaults to [] per version when not already an array", () => {
  // Newest-first input (timestamps descending) so the oldest-first migration
  // check doesn't reorder these before the assertions below.
  const out = normalizeRaw({ id: 24, versions: [{ timestamp: 2, linkIds: [5, 6] }, { timestamp: 1 }] });
  assert.deepEqual(out.versions[0].linkIds, [5, 6]);
  assert.deepEqual(out.versions[1].linkIds, []);
});

await test("each version's classification falls back to the entity's classification", () => {
  const out = normalizeRaw({
    id: 25,
    classification: "regulated",
    versions: [{ timestamp: 2, classification: "  Public  " }, { timestamp: 1 }],
  });
  assert.equal(out.versions[0].classification, "public");
  assert.equal(out.versions[1].classification, "regulated");
});

// ── normalizeRaw — legacy oldest-first migration ────────────────────────────
console.log("\n── normalizeRaw: legacy oldest-first version array is reversed ───");

await test("an oldest-first versions array is detected and reversed to newest-first", () => {
  const out = normalizeRaw({
    id: 30,
    versions: [{ timestamp: 100 }, { timestamp: 200 }], // oldest-first input
  });
  assert.deepEqual(out.versions.map(v => v.timestamp), [200, 100], "must end up newest-first");
});

await test("an already newest-first versions array is left in order", () => {
  const out = normalizeRaw({
    id: 31,
    versions: [{ timestamp: 200 }, { timestamp: 100 }], // already newest-first
  });
  assert.deepEqual(out.versions.map(v => v.timestamp), [200, 100]);
});

await test("equal-timestamp versions are not reversed (strict < only)", () => {
  const out = normalizeRaw({
    id: 32,
    versions: [{ timestamp: 100, note: "first" }, { timestamp: 100, note: "second" }],
  });
  assert.deepEqual(out.versions.map(v => v.note), ["first", "second"]);
});

await test("a single-element versions array is untouched by the reversal check", () => {
  const out = normalizeRaw({ id: 33, versions: [{ timestamp: 100 }] });
  assert.equal(out.versions.length, 1);
});

// ── normalizeRaw — unicode ──────────────────────────────────────────────────
console.log("\n── normalizeRaw: unicode content passes through intact ───────────");

await test("unicode text, tags, and workspaceId survive normalization", () => {
  const out = normalizeRaw({
    id: 40,
    text: "こんにちは世界 🎉",
    tags: ["日本語", "emoji🔥"],
    workspaceId: "  日本語ワークスペース  ",
    versions: [],
  });
  assert.equal(out.text, "こんにちは世界 🎉");
  assert.deepEqual(out.tags, ["日本語", "emoji🔥"]);
  assert.equal(out.workspaceId, "日本語ワークスペース");
});

// ── normalizeRaw — determinism ──────────────────────────────────────────────
console.log("\n── normalizeRaw: determinism ──────────────────────────────────────");

await test("normalizing two structurally-identical (but distinct) raw objects yields identical results", () => {
  const build = () => ({
    id: 99,
    text: "duplicate content",
    tags: ["x", "y"],
    versions: [{ timestamp: 1000 }, { timestamp: 500 }],
  });
  const outA = normalizeRaw(build());
  const outB = normalizeRaw(build());

  // Sets don't compare structurally with deepEqual/JSON.stringify; compare
  // everything else, plus the links contents explicitly.
  const strip = (o) => { const { links, ...rest } = o; return rest; };
  assert.deepEqual(strip(outA), strip(outB));
  assert.deepEqual([...outA.links], [...outB.links]);
});

await test("re-running normalizeRaw on an already-normalized entity is idempotent", () => {
  const raw = { id: 100, text: "idempotent check", versions: [{ timestamp: 1 }, { timestamp: 2 }] };
  const once  = normalizeRaw(raw);
  const dump1 = JSON.stringify({ ...once, links: [...once.links] });
  const twice = normalizeRaw(once); // same object, ran again
  const dump2 = JSON.stringify({ ...twice, links: [...twice.links] });
  assert.equal(dump1, dump2, "a second pass over already-normalized data must not change it");
});

// ── Results ──────────────────────────────────────────────────────────────────
console.log(`\n${"─".repeat(60)}`);
const total = passed + failed;
console.log(`  ${passed}/${total} passed${failed ? `  (${failed} failed)` : " ✅"}`);
console.log(`${"─".repeat(60)}\n`);
process.exit(failed > 0 ? 1 : 0);

})();
